content/post/kmcuda4.md

```diff
@@ -0,0 +1,489 @@
+may end up with the wrong device. This is not cool at all and I think
+the weakest part of the whole CUDA runtime API. Setting the device number for
+every API function would be much better, and it would cure all the
+thread safety issues which are currently ugly solved by the thread local
+context.
+
+You shouldn't mix kernel calls with peer to peer memory exchanges,
+because `cudaMemcpyPeerAsync()` requires **both** device pipelines
+to become ready. In other words, the following will serialize the computation:
+
+```c
+for each device {
+  kernel<<<...>>>(...);
+  for all other devices {
+    // wrong - will block other device's stream 
+    cudaMemcpyPeerAsync(this device, other device);
+  }
+}
+```
+
+But this will work:
+
+```c
+for each device {
+  kernel<<<...>>>(...);
+}
+// all pipelines are loaded now
+for each device {
+  for all other devices {
+    cudaMemcpyPeerAsync(this device, other device);
+  }
+}
+```
+
+I wrapped all the repeating boilerplate code in macros. While I am DRY
+now, I clearly stepped into a macro hell. Here is an example. An innocent
+line which allocates ccounts (centroid assignments counters
+written on stage 1 and read on stage 2)...
+
+```
+CUMALLOC(device_ccounts, clusters_size);
+``` 
+
+...is actually...
+
+```c++
+do {
+  do {
+    for (int dev : devs) {
+      cudaSetDevice(dev);
+      do {
+        void* __ptr;
+        do {
+          auto __res = cudaMalloc(
+              &__ptr,
+              (clusters_size) *
+                  sizeof(std::remove_reference<decltype(
+                             device_ccounts)>::type::value_type::element_type));
+          if (__res != 0) {
+            do {
+              if (verbosity > 1) {
+                printf("%s\n",
+                       "cudaMalloc( &__ptr, (clusters_size) * "
+                       "sizeof(std::remove_reference<decltype(device_ccounts)>:"
+                       ":type::value_type ::element_type))");
+              }
+            } while (false);
+            do {
+              if (verbosity > 0) {
+                printf("%s:%d -> %s\n", "_file_name_", 301,
+                       cudaGetErrorString(__res));
+              }
+            } while (false);
+            do {
+              if (verbosity > 0) {
+                printf(
+                    "failed to allocate %zu bytes for "
+                    "device_ccounts"
+                    "\n",
+                    static_cast<size_t>(clusters_size));
+              }
+            } while (false);
+            return kmcudaMemoryAllocationFailure;
+          }
+        } while (false);
+        (device_ccounts)
+            .emplace_back(
+                reinterpret_cast<std::remove_reference<decltype(
+                    device_ccounts)>::type::value_type::element_type*>(__ptr));
+      } while (false);
+    }
+  } while (false);
+} while (false)
+```
+
+And this is only one tiny example.
+I am really sorry what I have done. However, I don't
```

Comment:
This sentence doesn't sound as correct English to me, but I don't know how to improve it. Maybe something along the lines of: "I regret having end up using macros like this"...